opentelemetry-api/src/opentelemetry/metrics/_internal/__init__.py

```diff
@@ -218,38 +232,58 @@ def schema_url(self) -> Optional[str]:
         """
         return self._schema_url
 
-    def _is_instrument_registered(
-        self, name: str, type_: type, unit: str, description: str
-    ) -> Tuple[bool, str]:
+    def _register_instrument(
+        self,
+        name: str,
+        type_: type,
+        unit: str,
+        description: str,
+        advisory: Optional[MetricsInstrumentAdvisory] = None,
+    ) -> Tuple[str, _InstrumentRegistrationStatus]:
         """
-        Check if an instrument with the same name, type, unit and description
-        has been registered already.
-
-        Returns a tuple. The first value is `True` if the instrument has been
-        registered already, `False` otherwise. The second value is the
-        instrument id.
+        Register an instrument with the name, type, unit and description as
+        identifying keys and the advisory as value.
+
+        Returns a tuple. The first value is the instrument id.
+        The second value is an `_InstrumentRegistrationStatus` where
+        `already_registered` is `True` if the instrument has been registered
+        already.
+        If `conflict` is set to True the `current_advisory` attribute contains
+        the registered instrument advisory.
         """
 
         instrument_id = ",".join(
             [name.strip().lower(), type_.__name__, unit, description]
         )
 
-        result = False
+        already_registered, conflict = False, False
+        current_advisory = None
 
         with self._instrument_ids_lock:
-            if instrument_id in self._instrument_ids:
-                result = True
+            # we are not using get because None is a valid value
+            already_registered = instrument_id in self._instrument_ids
+            if already_registered:
+                current_advisory = self._instrument_ids[instrument_id]
+                conflict = current_advisory != advisory
             else:
-                self._instrument_ids.add(instrument_id)
-
-        return (result, instrument_id)
+                self._instrument_ids[instrument_id] = advisory
+
+        return (
+            instrument_id,
+            _InstrumentRegistrationStatus(
+                already_registered=already_registered,
+                conflict=conflict,
+                current_advisory=current_advisory,
+            ),
+        )
 
     @abstractmethod
     def create_counter(
         self,
         name: str,
         unit: str = "",
         description: str = "",
+        advisory: None = None,
```

Comment:
Wdyt of adding doing this?
```suggestion
        *,
        advisory: None = None,
```

to force any new arguments to be passed by name explicitly
- getting to having a lot of arguments here and IMO it is less error prone for users
- that would allow us to reorder the argument list if needed in the future

if we also added a trailing `**kwargs`, we could even remove parameters in the future if they get deprecated. Maybe lets discuss in the SIG but wanted to get your thoughts.

Reply:
I'm fine with that

Reply:
Forgot to bring it at the SIG but pushed this since:
- it has many advantages
- we can revert it if it's a problem

Reply:
> * we can revert it if it's a problem

true